lib/index.ts

```diff
@@ -0,0 +1,295 @@
+import fs = require('fs');
+import path = require('path-posix');
+import symlinkOrCopy = require('symlink-or-copy');
+import Logger = require('heimdalljs-logger');
+
+import Entry, { DefaultEntry, BaseEntry } from './entry';
+import {
+  sortAndExpand,
+  validateSortedUnique
+} from './util';
+
+declare namespace FSTree {
+  export type Operand = 'unlink' | 'rmdir' | 'create' | 'change' | 'mkdir';
+  export type Operation = [Operand, string, DefaultEntry ] | [ Operand, string];
+  export type Patch = Operation[];
+  export type IEntry = import('./entry').DefaultEntry;
+
+  export interface Options {
+    entries?: BaseEntry[],
+    sortAndExpand?: boolean
+  }
+
+  export interface StaticOptions {
+    sortAndExpand?: boolean
+  }
+
+  interface PatchDelegate {
+    unlink?(inputPath: string, outputPath: string, relativePath: string) : void;
+    rmdir?(inputPath: string, outputPath: string, relativePath: string) : void;
+    mkdir?(inputPath: string, outputPath: string, relativePath: string) : void;
+    change?(inputPath: string, outputPath: string, relativePath: string) : void;
+    create?(inputPath: string, outputPath: string, relativePath: string) : void;
+  }
+}
+
+const logger = Logger('fs-tree-diff:');
+const ARBITRARY_START_OF_TIME = 0;
+const DEFAULT_DELEGATE: FSTree.PatchDelegate  = {
+  unlink(inputPath: string, outputPath: string, relativePath: string) {
+    try {
+      fs.unlinkSync(outputPath);
+    } catch (e) {
+      if (typeof e === 'object' && e !== null && e.code === 'ENOENT') {
+        return;
+      }
+      throw e;
+    }
+  },
+  rmdir(inputPath: string, outputPath: string, relativePath: string) {
+    fs.rmdirSync(outputPath)
+  },
+  mkdir(inputPath: string, outputPath: string, relativePath: string) {
+    fs.mkdirSync(outputPath);
+  },
+  change(inputPath: string, outputPath: string, relativePath: string) {
+    // We no-op if the platform can symlink, because we assume the output path
+    // is already linked via a prior create operation.
+    if (symlinkOrCopy.canSymlink) {
+      return;
+    }
+
+    fs.unlinkSync(outputPath);
+    symlinkOrCopy.sync(inputPath, outputPath);
+  },
+  create(inputPath: string, outputPath: string, relativePath: string) {
+    symlinkOrCopy.sync(inputPath, outputPath);
+  }
+};
+
+class FSTree<T extends BaseEntry> {
```

Comment:
You might consider making the default entry type the default when you don't provide the generic type. `class FSTree<T extends BaseEntry = DefaultEntry>`

Reply:
I did not realize this was a thing. Cool!